hashjoin.go

```diff
@@ -0,0 +1,393 @@
+package exectoy
+
+// todo(changangela) hash table should be able to rehash or determine bucket size dynamically
+const hashTableBucketSize = 1 << 16
+
+type hashTableInt struct {
+	first []int
+	next []int
+
+	keys intColumn
+	values []intColumn
+
+	nOutCols int
+	count int
+	bucketSize int
+
+	// allocated represents the reserved number of keys that the hashTableInt
+	// can store in its current state.
+	allocated int
+}
+
+func makeHashTableInt(bucketSize int, nOutCols int) *hashTableInt {
+	values := make([]intColumn, nOutCols)
+	for i := 0; i < nOutCols; i++ {
+		values[i] = make(intColumn, 1)
+	}
+
+	hashTable := &hashTableInt{
+		first: make([]int, bucketSize),
+		next: make([]int, 1),
+		keys: make([]int, 1),
+		values: values,
+		nOutCols: nOutCols,
+		bucketSize: bucketSize,
+		allocated: 1,
+	}
+
+	return hashTable
+}
+
+func (hashTable *hashTableInt) hashInt(key int) int {
+	// todo(changangela) replace with a better hash function or use a library
+	return key % hashTable.bucketSize
+}
+
+func (hashTable *hashTableInt) grow(amount int) {
+	hashTable.next = append(hashTable.next, make([]int, amount)...)
+	hashTable.keys = append(hashTable.keys, make(intColumn, amount)...)
```

Comment:
This may cause multiple allocations. Depending on whether you want fine-grained control over how much the slice will grow, I think the right way to do this is to allocate a new slice if the old's slice capacity is too small to fit the new amount, and copy the old slice into the new slice...

I hear Go 1.11 is more optimal for this case (https://go-review.googlesource.com/c/go/+/109517) but unfortunately we're still on 1.10 for other reasons. If this is your bottleneck i'd consider changing it, otherwise I guess we can leave for now.

Reply:
Yeah that's true, I'll change it.

Reply:
Yeah that makes sense, the implementation's a little different with the `ColVec` stuff and different types, which we can discuss later.